signalling/src/handlers/mod.rs

```diff
@@ -0,0 +1,1081 @@
+use webrtcsink_protocol as p;
+use anyhow::{anyhow, Error};
+use futures::prelude::*;
+use futures::ready;
+use pin_project_lite::pin_project;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tracing::{info, instrument};
+
+pin_project! {
+    #[must_use = "streams do nothing unless polled"]
+    pub struct Handler {
+        #[pin]
+        stream: Pin<Box<dyn Stream<Item=(String, Option<p::IncomingMessage>)> + Send>>,
+        items: VecDeque<(String, p::OutgoingMessage)>,
+        producers: HashMap<String, HashSet<String>>,
+        consumers: HashMap<String, Option<String>>,
+        listeners: HashSet<String>,
+    }
+}
+
+impl Handler {
+    #[instrument(level = "debug", skip(stream))]
+    /// Create a handler. When extending the protocol, pass in an
+    /// always pending stream, then call handle() and consume items until
+    /// pending
```

Comment:
I think technically more correct here would be not just a `Pending` stream, but you would also remember the `Waker` from the `Context` in `poll_next()` if there is currently no item available and then wake up the `Waker` whenever an outgoing message becomes available (i.e. `handle()` was called and produced something).

Reply:
correct, handle was unpubbed however so obsolete (at least for now)